Makie visualization for general meshes

jlchan opened the pull request:
Makie recipes implemented. This PR introduces `trixi_pcolor, trixi_pcolor!` and `trixi_wireframe, trixi_wireframe!` which plot pseudocolor and wireframe plots, respectively. The function names are due to Makie's plot recipe system, which creates them automatically. However, we could also overload Makie's existing plotting functions `mesh` and `lines` if that's preferable. 

The plotting functions just take `sol::TrixiODESolution` as an input. The internals could probably be optimized some, but it works OK as-is. Adding the following lines 
```julia
using GLMakie
trixi_pcolor(sol,1)
trixi_wireframe!(sol,1,color=:white) 
Makie.current_figure()
```
to the elixir from @andrewwinters5000 https://github.com/trixi-framework/Trixi.jl/blob/bea6ab9c9e68f944a1d7cfedcae826d42c1f1f8b/src/visualization/test_vis/ini_bara_ostrukturerad.jl generates the figure
![image](https://user-images.githubusercontent.com/1156048/119928626-89c37680-bf41-11eb-95d5-7203ff94c5cf.png)

One major downside is that Makie is not a light library, so Trixi's compile time increases. This should be fixed when the devs migrate some of the recipe functionality from Makie to MakieCore (see https://github.com/JuliaPlots/Makie.jl/issues/996).

Closes #754

sloede reviewed (changes requested):
Sorry for taking to long to review this. I have not really been able to review the `makie_visualization.jl` file properly since I don't follow the overall structure. Maybe we could have a short meeting about it where you walk me through the process? Then I think I would be able to give it a better shot...

jlchan commented on src/Trixi.jl:
```diff
@@ -171,16 +171,26 @@ export convergence_test, jacobian_fd, jacobian_ad_forward, linear_structure
+                
+  @require Makie="ee78f7c6-11fb-53f2-987a-cfe4a2b5a57a" begin
+    include("visualization/makie_visualization.jl")
+    using .Makie: Makie, Attributes, mesh, mesh!, plot!, @recipe, lines!, @lift
```
I think that makes sense. I'll go ahead and remove the conflicting names.

jlchan commented on src/visualization/test_vis/ini_bara_ostrukturerad.jl:
```diff
@@ -0,0 +1,87 @@
+
```
Ah, sorry - didn't realize I'd committed this. I was going to remove this until I figured out a good way to write a visualization test.

jlchan replied:
Removed - will commit soon.

jlchan commented:
> Sorry for taking to long to review this. I have not really been able to review the `makie_visualization.jl` file properly since I don't follow the overall structure. Maybe we could have a short meeting about it where you walk me through the process? Then I think I would be able to give it a better shot...

No worries, and sure - I'd be happy to. I'll try to clean things up tonight too so it's easier to review.

jlchan commented on src/visualization/test_vis/Project.toml:
```diff
@@ -0,0 +1,7 @@
+[deps]
+GLMakie = "e9467ef8-e4e7-5192-8a1a-b1aee30e663a"
+GeometryBasics = "5c1252a2-5f33-56bf-86c9-59e7332b4326"
+NodesAndModes = "7aca2e03-f7e2-4192-9ec8-f4ca66d597fb"
+OrdinaryDiffEq = "1dea7af3-3e70-54e6-95c3-0bf5283fa5ed"
+Triangulate = "f7e6ffb2-c36d-4f8f-a77e-16e897189344"
+Trixi = "a7f1ee26-1774-49b1-8366-f1abc58fbfcb"
```
I've removed the `test_vis` folder from this PR

jlchan commented:
@sloede the last commit cleaned up `src/visualization/makie_visualization.jl` as follows:

- added more docstrings
- renamed some variables to be consistent with PlotData
- removed `trixi_pcolor` and `trixi_wireframe`, which can probably be replaced with simpler ["type recipes"](https://makie.juliaplots.org/stable/recipes.html#Type-recipes) later on.

FelipeSantillan commented:
I have tested what was added in this PR and I think the plots look really nice.
I'm having a small issue; when I create a plot there are some switches to (for example) enable/disable the mesh, but I am unable to interact with them. I'm not sure if I'm doing something wrong...  Maybe you can help me here @jlchan

jlchan commented:
Thanks! I can take a look in a few days. For now, I guess something broke with the last version of Makie￼...

jlchan commented:
> I have tested what was added in this PR and I think the plots look really nice.
> I'm having a small issue; when I create a plot there are some switches to (for example) enable/disable the mesh, but I am unable to interact with them. I'm not sure if I'm doing something wrong... Maybe you can help me here @jlchan

Works for me, but I have no way to test on other OS right now. What OS are you using, and do the switches just not respond when you press them?

FelipeSantillan commented:
> > I have tested what was added in this PR and I think the plots look really nice.
> > I'm having a small issue; when I create a plot there are some switches to (for example) enable/disable the mesh, but I am unable to interact with them. I'm not sure if I'm doing something wrong... Maybe you can help me here @jlchan
> 
> Works for me, but I have no way to test on other OS right now. What OS are you using, and do the switches just not respond when you press them?

I am using Windows 10. The plot appears to be a still image and do not respond. I also tried using Atom and VSCode and got the same result.

jlchan commented:
@FelipeSantillan did you open an issue with Makie on the example http://makie.juliaplots.org/dev/makielayout/toggle.html not working?

jlchan commented:
One question on this PR - I'm planning on adding Plots.jl visualization for `DGMulti` solvers using TriplotBase.jl + TriplotRecipes.jl. 

I'd like to reuse some of the stuff in this PR, but it doesn't sound ready to merge yet. Should I just create a new PR and copy the necessary code from this PR over?

ranocha commented:
> I'd like to reuse some of the stuff in this PR, but it doesn't sound ready to merge yet. Should I just create a new PR and copy the necessary code from this PR over?

Yes, that sounds reasonable to me.

jlchan commented:
Adding image for docs 
<img width="794" alt="Screen Shot 2021-08-02 at 2 42 08 PM" src="https://user-images.githubusercontent.com/1156048/127914785-bb87cff5-92b2-4e72-9ed1-9c04f4f81a73.png">

jlchan commented:
@ranocha The PR should be ready to review now. I'm considering also adding the ability to pass in `cons2prim` or other solution transformations, but would also be OK leaving that to a future PR.

ranocha commented on src/visualization/makie_visualization.jl:
````diff
@@ -0,0 +1,263 @@
+    return t[:, findall(has_volume)]
+end
+
+"""
+  function plotting_interpolation_matrix(dg; kwargs...)
+
+Interpolation matrix which maps discretization nodes to a set of plotting nodes.
+Defaults to the identity matrix of size `length(solver.basis.nodes)`, and interpolates
+to equispaced nodes for DGSEM (set by kwarg `nvisnodes` in the plotting function).
+
+Example:
+```julia
+A = plotting_interpolation_matrix(dg)
+A * dg.basis.nodes # => vector of nodes at which to plot the solution
+```
+"""
+# Note: we cannot use UniformScaling to define the interpolation matrix since we use it with `kron`
+# to define a multi-dimensional interpolation matrix later.
````
Same as above: If it's an internal function, the docstring should be converted into a comment

ranocha commented on src/visualization/makie_visualization.jl:
````diff
@@ -0,0 +1,263 @@
+    return polynomial_interpolation_matrix(dg.basis.nodes, LinRange(-1, 1, nvisnodes))
+end
+
+"""
+  function generate_plotting_triangulation(sol::TrixiODESolution, variable_to_plot::Int; nvisnodes=5)
+
+Generates a plotting triangulation which can be used by Makie's `mesh` function.
+
+Arguments:
+- sol: TrixiODESolution
+- variable_to_plot: index of solution field to plot
+
+Example:
+```julia
+tri = generate_plotting_triangulation(sol,1)
+Makie.mesh(tri,color=getindex.(tri.position,3))
+```
+The color is specified to be the z-value of the solution, which there does not seem to be a default solution for yet.
+"""
````
Same as above: If it's an internal function, the docstring should be converted into a comment

ranocha commented on src/visualization/makie_visualization.jl:
````diff
@@ -0,0 +1,263 @@
+```
+The color is specified to be the z-value of the solution, which there does not seem to be a default solution for yet.
+"""
+function generate_plotting_triangulation(sol::TrixiODESolution, variable_to_plot::Int; nvisnodes=10)
+
````
It looks like this is only working for 2D. Could you please add a corresponding `@assert` somewhere at the beginning of this method?

jlchan replied:
added `@assert ndims(mesh) == 2`. Does that sound OK?

ranocha commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,263 @@
+    coordinates_tmp = zeros(Float32, num_nodes, 3)
+    coordinates = zeros(Float32, n_plot_nodes, 3)
+    for element in eachelement(dg, cache)
+
```
```suggestion
```

ranocha commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,263 @@
+        for j in eachnode(dg), i in eachnode(dg)
+            u_node = get_node_vars(u, equations, dg, i, j, element)
+            xy_node = get_node_coords(cache.elements.node_coordinates, equations, dg, i, j, element)
+            coordinates_tmp[sk,1:2] .= xy_node
+            coordinates_tmp[sk,3] = u_node[variable_to_plot]
```
```suggestion
            coordinates_tmp[sk, 1:2] .= xy_node
            coordinates_tmp[sk, 3] = u_node[variable_to_plot]
```

ranocha commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,263 @@
+
+function generate_plotting_wireframe(sol::TrixiODESolution, variable_to_plot::Int; nvisnodes=5)
+
+    semi = sol.prob.p
```
Same comment on `@assert` as above

ranocha commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,263 @@
+    dg = semi.solver
+
+    num_nodes_1D = length(dg.basis.nodes)
+    num_nodes = num_nodes_1D^2
```
I guess this `2` here is something like `ndims(mesh)`? If so, could you please use that to make it more clear?

ranocha commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,263 @@
+
+    num_nodes_1D = length(dg.basis.nodes)
+    num_nodes = num_nodes_1D^2
+    num_elements = nelements(dg,cache)
```
```suggestion
    num_elements = nelements(dg, cache)
```

ranocha commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,263 @@
+
+    # reconstruct reference nodes (assumes 2D ordering on reference quad is x_i,y_j with i first).
+    r1D = dg.basis.nodes
+    r = vec([r1D[i] for i = 1:num_nodes_1D, j = 1:num_nodes_1D])
+    s = vec([r1D[j] for i = 1:num_nodes_1D, j = 1:num_nodes_1D])
```
```suggestion
    r = vec([r1D[i] for i in 1:num_nodes_1D, j in 1:num_nodes_1D])
    s = vec([r1D[j] for i in 1:num_nodes_1D, j in 1:num_nodes_1D])
```

ranocha commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,263 @@
+    # The resulting arrays are then reshaped so that xf, yf, sol_f are Matrix types of size
+    # (Number of face plotting nodes) x (Number of faces).
+    function face_first_reshape(x, num_nodes_1D, num_nodes, num_elements)
+        num_reference_faces = 4 # hardcoded for quadrilateral elements
```
`4 == 2 * ndims(mesh)`?

ranocha commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,263 @@
+    return Makie.Point.(xfp, yfp, ufp)
+end
+
+function trixi_plot(sol::TrixiODESolution; variable_to_plot_in = 1, nvisnodes=5)
+
+    semi = sol.prob.p
+    dg = semi.solver
```
Same comment on `@assert` as above

jlchan replied:
In `trixi_plot`, right?

ranocha replied:
Yes :+1:

ranocha reviewed (changes requested):
Thanks a lot, @jlchan! This looks already quite nice. Please find below some comments, questions, and minor bikeshedding suggestions.

jlchan commented on Project.toml:
```diff
@@ -28,6 +29,7 @@ StrideArrays = "d1fa6d79-ef01-42a6-86c9-f7c551f8593b"
 StructArrays = "09ab397b-f2b6-538f-b94a-2f83cf4a842a"
 SummationByPartsOperators = "9f78cca6-572e-554e-b819-917d2f1cf240"
 TimerOutputs = "a759f4b9-e2f1-59dc-863e-4aeb61b1ea8f"
+Triangulate = "f7e6ffb2-c36d-4f8f-a77e-16e897189344"
```
Good point. I'll go ahead and do that in this PR too.

jlchan commented on src/Trixi.jl:
```diff
@@ -213,6 +214,12 @@ function __init__()
   @require Plots="91a5bcdd-55d7-5caf-9e0b-520d859cae80" begin
     using .Plots: plot, plot!, savefig
   end
+
+  @require Makie="ee78f7c6-11fb-53f2-987a-cfe4a2b5a57a" begin
+    include("visualization/makie_visualization.jl")
+    using .Makie: Makie, Attributes, mesh, mesh!, plot!, @recipe, lines!, @lift
```
I'd prefer to import Plots routines and just have `using .Makie: Makie`. This would make it so that users could still use the Plots.jl recipes and type `plot(sol)` instead of `Plots.plot(sol)`. Does that sound OK?

ranocha replied:
This stuff is just inside Trixi.jl. Thus, users will always be able to do something like
```julia
using Plots
plot(sol)
```
Restricting the using statement only for Makie would also be okay with me, but it might be a bit more clear for people reading the visualization code when we just prefix the few plotting commands there with `Plots.`.

jlchan replied:
Ah, I see. Agreed, especially if we can overload `plot` in Makie as well.

jlchan commented on src/Trixi.jl:
```diff
@@ -213,6 +214,12 @@ function __init__()
+  @require Makie="ee78f7c6-11fb-53f2-987a-cfe4a2b5a57a" begin
+    include("visualization/makie_visualization.jl")
+    using .Makie: Makie, Attributes, mesh, mesh!, plot!, @recipe, lines!, @lift
+    export trixi_plot
```
They do allow user/type recipes. The type recipes hook into existing plot functions, and Makie.jl doesn't use `plot` as an actual plot signature. My understanding was that the user recipes let you define new plot commands (e.g., `myplot(customType)`, but I don't think the recipe enables `Makie.plot(customType)`. 

I'll check on this - it would be nice to have a unified `plot(sol)` interface.

jlchan replied:
Looks like we can overload `Makie.plot` after all. Let me try to add a user recipe for `TrixiODESolution` in this PR.

ranocha replied:
Sounds interesting :+1: Using the setup you currently have is also okay for me - I forgot that Makie uses a different interface than other plotting packages

jlchan commented:
A few updates:

1. I've found a way to make heatmap plots in Makie
2. we can indeed use custom recipes in Makie, e.g., `Makie.plot(MyCustomType)`

Thus, we can recreate all of Trixi's main Plots.jl visualization capabilities using Makie! 

I have another PR https://github.com/trixi-framework/Trixi.jl/pull/755 which adds Plots.jl visualization to `DGMulti`. This PR doesn't do anything with `DGMulti`, and also doesn't utilize the same Plots.jl API. Thus, I'd suggest the following approach:

1. Add Makie heatmap capabilities 
2. Add `UnstructuredPlotData2D` (a generalization of `PlotData2D`) to this PR
3. Merge this PR into `dev`
4. Merge `dev` into https://github.com/trixi-framework/Trixi.jl/pull/755
5. Use `UnstructuredPlotData2D` in https://github.com/trixi-framework/Trixi.jl/pull/755 for Plots.jl support
6. Merge into `dev`, review, then merge into `main`

@ranocha how does that sound? May also be worth checking with sloede (will tag him when he's back from vacation) to review the plot API too, since this builds so heavily on it.

ranocha commented:
> Thus, we can recreate all of Trixi's main Plots.jl visualization capabilities using Makie!

This is great news :+1: 

> How does the following approach sound:
> 
> 1. Add Makie heatmap capabilities
> 2. Add `UnstructuredPlotData2D` (a generalization of `PlotData2D`) to this PR
> 3. Merge this PR into `dev` or `main`
> 4. Merge `dev` or `main` into [WIP: Plots.jl recipes for DGMulti solvers #755](https://github.com/trixi-framework/Trixi.jl/pull/755)
> 5. Adapt `UnstructuredPlotData2D` in [WIP: Plots.jl recipes for DGMulti solvers #755](https://github.com/trixi-framework/Trixi.jl/pull/755) for Plots.jl support

This sounds good to me. From my point of view, there is no need to merge into `dev`, we can directly target `main` for this.

jlchan commented:
There are some [limitations in Makie's recipes](https://discourse.julialang.org/t/accessing-axis-in-makie-plot-recipes/66006), which means that `Makie.plot(::PlotDataSeries2D)` is missing a title, the axes aren't tight, and I haven't yet figured out how to toggle grid lines using the old Plots.jl API. However, we can do the following in Makie now:

![image](https://user-images.githubusercontent.com/1156048/128659358-0ff7e14d-a8b1-4ee8-837f-92eccc94557c.png)

jlchan commented:
OK - apologies that this PR has grown so much in scope. I'm happy to create a new PR if that'll be easier review. 
Most changes are in `makie_visualization.jl`. Some type changes were made in `plot_recipes.jl` in order to reuse `PlotDataSeries2D`. 

For now, a summary of Makie plotting capabilities:

1. There is a heatmap-type visualization, accessible using the old Plots.jl API (e.g., `plot(sol)` or `plot(PlotData2D(sol)["rho"])`). The main difference is that mesh plotting is toggled using a keyword argument instead of `plot!(getmesh(pd))`.
![image](https://user-images.githubusercontent.com/1156048/129269969-cd5ae4de-8b5d-4810-bcba-e2fbcbc907ce.png)

2. There is an interactive surface plot using `trixi_plot(sol)` where you can rotate and zoom the solution.

jlchan commented:
From Codecov, it looks like there's some slice-plotting functionality that I erased when I removed some conflicting tests, hence the failing project coverage test.

ranocha commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,348 @@
+  fig
+end
+
+function Makie.plot!(fig, pd::UnstructuredPlotData2D;
+                     plot_mesh = true)
```
Is it possible to modify the figure returned by this function afterwards, e.g., to remove the title again when considering to use this for a publication?

jlchan replied:
For this specific function, I think so - you can modify the titles afterwards using the Makie figure object `fig` returned by `Makie.plot`. 

For the plot recipe triggered by `Makie.plot(pd["rho"])`, I'm not sure, since there's no explicit access to Makie's layout objects (figure/axis). Let me check on Slack/Discourse first though.

jlchan replied:
I've modified the custom plot routine to return the figure and axes handles - you can modify the titles using the axis handles.

ranocha commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,348 @@
+    cols = ceil(Int, sqrt(length(pd)))
+    rows = ceil(Int, length(pd)/cols)
+  end
+
```
Would it be possible to add the axis labels by default as we do for Plots.jl?

jlchan replied:
What do you mean?

jlchan replied:
But there don't seem to be axis labels on the Plots recipes, at least on the doc images https://trixi-framework.github.io/Trixi.jl/dev/visualization/

ranocha replied:
Yeah, that figure is old and we should update it (#788). In newer versions, we use axis labels, see https://github.com/trixi-framework/Trixi.jl/blob/main/src/visualization/plot_recipes.jl#L309-L310 and https://github.com/trixi-framework/Trixi.jl/blob/main/src/visualization/plot_recipes.jl#L281-L291

sloede replied:
Tracked in #792.

ranocha commented on test/test_visualization.jl:
```diff
@@ -19,7 +20,6 @@ isdir(outdir) && rm(outdir, recursive=true)
   test_examples_2d = Dict(
     "TreeMesh" => ("tree_2d_dgsem", "elixir_euler_blast_wave_amr.jl"),
     "StructuredMesh" => ("structured_2d_dgsem", "elixir_euler_source_terms_waving_flag.jl"),
-    "UnstructuredMesh2D" => ("unstructured_2d_dgsem", "elixir_euler_wall_bc.jl")
```
Why did you remove this one?

jlchan replied:
This PR causes the old recipes to fail because we're pirating the `PlotData2D(...)` API. However, the old recipes for `UnstructuredMesh2D` were just scatter plots, and were essentially a placeholder until more mature functionality was introduced. 

I'm not sure of the best way around this conflict. We break Plots.jl functionality for `UnstructuredMesh2D` in this PR but won't replace it until #755 or #785. Perhaps we could merge this PR to `dev`, then merge #755 and consolidate them inside `dev`?

sloede replied:
If UnstructuredMesh2D really only used scatter plots, I don't think we lose anything by breaking support for it. Especially since we have support for proper plotting coming in from @FelipeSantillan's PR. Thus unless @andrewwinters5000 objects, I'd say out the door with it!

jlchan replied:
I was worried previously that there was some slice support in `UnstructuredMesh2D` that I was erasing, but it looks like that's not the case. In other words, this is a direct replacement of plotting capabilities.

ranocha replied:
Okay, let's go ahead :+1:

ranocha reviewed (changes requested):
Thanks a lot! This is a really nice feature :+1: Please find below some comments and questions.

jlchan commented on src/visualization/makie_vis_template.jl:
```diff
@@ -0,0 +1,26 @@
+struct MyType end
+@Makie.recipe(MyPlot, mytype) do scene
+  Makie.Theme(;)
+end
+function Makie.plot!(myplot::MyPlot{Tuple{<:MyType}})
+  x = LinRange(-1,1,100)
+  Makie.lines!(myplot, sin.(pi*x))
+  Makie.plot!(myplot, cos.(pi*x))
+  myplot
+end
+Makie.plottype(::MyType) = MyPlot # makes plot(::MyType) = myplot(::MyType)
+
+
+struct MyOtherType end
+function Makie.plot(x::MyOtherType)
+  fig = Makie.Figure()
+  axes = [Axis(fig[i, j]) for j = 1:3, i = 1:2]
+  for ax in axes
+    myplot!(ax, MyType())
+  end
+  fig
+end
+
+# 1. create recipe for UnstructuredPlotDataSeries2D
+# 2. use type conversion to overload Makie.plot(UnstructuredPlotDataSeries2D)
+# 3. overload Makie.plot(UnstructuredPlotData2D) to create layout. check if this lets you use Makie.plot() too?
```
Whoops - yes, I meant to remove this once the PR was ready for review. I'll delete it soon.

jlchan commented on docs/src/visualization.md:
````diff
@@ -398,3 +398,50 @@ to convert all solution files. The default is to generate a PNG file for each
+julia> using GLMakie
+```
+
+To visualize the solution and mesh with a heatmap-type plot, simply run
+```julia
+julia> plot(sol)
+```
````
Correct; however, it also works for `TreeMesh` if we remove specialization. However, I didn't want to overwrite `TreeMesh`'s Plots.jl visualization capabilities until we've had a chance to discuss consolidating visualization capabilities between Plots.jl, this PR, #755, and now #785.

jlchan commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,348 @@
+  @unpack xf, yf, uf = pd
+
+  if set_z_coordinate_zero==true
+    sol_f = zeros(size(xf)) # plot 2d surface by setting z coordinate to zero
+  else
+    sol_f = StructArrays.component(uf, variable_id)
```
Good idea! I'll add that.

jlchan commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,348 @@
+- solution_variables: either `nothing` or a variable transformation function (e.g., `cons2prim`)
+- nvisnodes: number of visualization nodes per dimension
+"""
+function trixi_plot(sol::TrixiODESolution;
+                    solution_variables=nothing, nvisnodes=5, variable_to_plot_in = 1)
```
Correct; I'll add that to the docstrings.

jlchan commented on src/visualization/plot_recipes.jl:
```diff
@@ -198,10 +198,21 @@ function _get_orientations(mesh, slice)
+  xf::Matrix{Tv}
+  yf::Matrix{Tv}
+  uf::FaceSolutionType
+  variable_names::VariableNames
```
Good to know!

jlchan commented on src/visualization/plot_recipes.jl:
```diff
@@ -198,10 +198,21 @@ function _get_orientations(mesh, slice)
 #
 # Note: This is an experimental feature and may be changed in future releases without notice.
-struct PlotDataSeries2D{PD<:PlotData2D}
+struct PlotDataSeries2D{PD<:Union{PlotData2D, UnstructuredPlotData2D}}
```
I was also thinking about this - I'll open up a new issue for it. I believe `PlotData2D` works for `TreeMesh`, `StructuredMesh`, and `UnstructuredMesh2D`, but the support for `UnstructuredMesh2D` is just a placeholder (a nodal scatter plot).

jlchan replied:
https://github.com/trixi-framework/Trixi.jl/issues/786

jlchan commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,348 @@
+    rows = 1
+  else
+    cols = ceil(Int, sqrt(length(pd)))
+    rows = ceil(Int, length(pd)/cols)
```
I think it is - this was copied this from the old Plots.jl code, which according to @sloede was written fairly early in Trixi.

jlchan commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,348 @@
+  ufp = StructArray{SVector{nvars, uEltype}}(map(reshape_and_interpolate, StructArrays.components(u)))
+
+  # convert variables based on solution_variables mapping
+  solution_variables_ = digest_solution_variables(equations, solution_variables)
+  for (i, u_i) in enumerate(uplot)
+    uplot[i] = solution_variables_(u_i, equations)
+  end
+  for (i, u_i) in enumerate(ufp)
+    ufp[i] = solution_variables_(u_i, equations)
+  end
+
+  variable_names = SVector(varnames(solution_variables_, equations))
```
Something like this?

```julia
  solution_variables_ = digest_solution_variables(equations, solution_variables)
  variable_names = SVector(varnames(solution_variables_, equations))

  transform_to_solution_variables!(uplot, solution_variables_, equations)
  transform_to_solution_variables!(ufp, solution_variables_, equations)
```

ranocha replied:
Looks good to me :+1:

sloede replied:
Seems to have been fixed by one of the commits already.

jlchan commented:
@sloede I've added some docs and changes to address some of the concerns @ranocha raised. It should be ready for another round of reviews.

jlchan commented on test/test_visualization.jl:
```diff
@@ -3,6 +3,7 @@ module TestVisualization
 using Test
 using Trixi
 using Plots
+using Makie
```
Oops - this should be GLMakie. That's why the test is failing.

I'll wait until @sloede is done with his review before updating.

sloede commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,373 @@
+
+"""
+    trixi_plot(sol::TrixiODESolution;
+               solution_variables=nothing, nvisnodes=5, variable_to_plot_in = 1)
```
Why not default to `nvisnodes=nothing` and have it be determined automatically?

jlchan replied:
Hm - `nvisnodes=nothing` doesn't work, and I wanted to provide the option to refine the visualization resolution. I've added a `default_nvisnodes(sol)` function which just returns `2*length(dg.basis.nodes)`.

sloede commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,373 @@
+
+"""
+    trixi_plot(sol::TrixiODESolution;
+               solution_variables=nothing, nvisnodes=5, variable_to_plot_in = 1)
```
Why specify the variable? Would it be possible to use a drop down to also allow variable selection?

jlchan replied:
It's just the default variable that `iplot` starts on. There is actually a drop down menu for variable selection. 

I figure that if someone wants to run `iplot` repeatedly and wanted to skip the dropdown menu step, they could specify `variable_to_plot_in` to bypass it.

sloede replied:
Good thinking - thanks!

sloede commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,373 @@
+end
+
+"""
+    trixi_plot(sol::TrixiODESolution;
```
I like `trixi_plot` since it's consistent with `trixi_include` or `trixi_test`. However, it seems to be a very generic name for a rather specific result (i.e., a very specific way of plotting), and also suggests that it somehow replaces a "standard" `plot` command (like it does for `trixi_include` and `trixi_test`). What about just using `plot_interactive` or `iplot` (in this case I'd argue that brevity might overrule descriptiveness)? Note that I do not have a fixed opinion or an optimal solution, it just feels slightly off as it is right now

jlchan replied:
I like `iplot` - especially since I don't like the idea of having to type `plot_interactive` each time...

ranocha replied:
It's only one character more (`iplot`: 5, `plot_` + TAB: 6). But I'm fine with `iplot` since that's what I do when calling this

sloede commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,373 @@
+
+  pd = PlotData2D(sol; solution_variables=solution_variables, nvisnodes=nvisnodes)
+
+  @unpack variable_names = pd
+
+  semi = sol.prob.p
+  @unpack mesh = semi
```
IIRC, one is supposed to the MESC extraction function for this purpose:
```suggestion
  @unpack variable_names = pd
  mesh, _, _, _ = mesh_equations_solver_cache(sol.prob.p)
```

ranocha replied:
Yeah, this will make it possible to use it for coupled Euler with self-gravity :+1:

jlchan replied:
Sweet!

sloede commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,373 @@
+
+# ================== new Makie plot recipes ====================
+
+@Makie.recipe(TrixiHeatmap, plot_data_series) do scene
+  Makie.Theme(;)
+end
```
I have zero understanding what this method does - obviously some black behind-the-scenes magic. Could you add a short comment telling what this does or why it's necessary?

jlchan replied:
I've added some comments, but much of it is just Makie black magic. Hope it's clearer now!

sloede replied:
Thanks!

sloede commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,373 @@
+  return Makie.plot(PlotData2D(sol, solution_variables=solution_variables); plot_mesh=plot_mesh)
+end
+
+# convenience struct for editing Makie plots after they're created.
+struct FigAxes{Axes}
+  fig::Makie.Figure
+  axes::Axes
+end
+
+# for "quiet" return arguments to Makie.plot(::TrixiODESolution) and
+# Makie.plot(::UnstructuredPlotData2D)
+Base.show(io::IO, fa::FigAxes) = nothing
```
Isn't this some sort of type piracy?

jlchan replied:
It shouldn't be since we define FigureAndAxes?

sloede replied:
You're right, now I understand how it's used.

sloede commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,373 @@
+end
+
+# convenience struct for editing Makie plots after they're created.
+struct FigAxes{Axes}
```
I'd call this `FigureAndAxes` to make it more descriptive and to avoid the misunderstanding "axes of the figure" (that's how I understood `FixAxes` when I first sawy it). Similarly, use `figure_and_axes` for the variables. The long name shouldn't be an issue since it's just used internally for convenience. OTOH, it's not a big issue for the same reason, so if you do not like this suggestion, it's ok to leave it as it is.

sloede reviewed (changes requested):
Great work - thanks for adding this! I am really looking forward to getting this merged such that @FelipeSantillan can continue extending support to other mesh types!

Please find some comments and suggestions; feel free to ask if anyting is unclear. As usual, everything is just a suggestion.

jlchan commented on docs/src/visualization.md:
````diff
@@ -398,3 +398,59 @@ to convert all solution files. The default is to generate a PNG file for each
+julia> plot(sol)
+```
+!!! note
+    Both Makie.jl and Plots.jl export `plot`, so if you load both librarys, you will have to
````
Wow, how did I miss this as a native English speaker?

sloede replied:
This is marked as resolved but not fixed in the sources. Did you, by any chance, hit `Add  suggestion to batch` and then either "Resolve conversation" or replied using the "Submit review" functionality? If you did, this will unremove the suggestion and not apply the change. Is there any chance this applies also to other comments I made?

jlchan replied:
That's exactly what I did...

I'm checking over each comment now in case I missed something.

jlchan commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,373 @@
+      ids = @view t[:, i]
+      x_points = @view triout.pointlist[1, ids]
+      y_points = @view triout.pointlist[2, ids]
+      area = compute_triangle_area(zip(x_points, y_points))
```
Oh yeah, I think this was replacing a broadcast which needed the zip. Will remove soon.

jlchan replied:
No, turns out I needed the zip too since `x_points` is an array.

sloede replied:
OK!

jlchan commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,373 @@
+  return kron(Vp1D, Vp1D)
+end
+
+function reference_node_coordinates_2d(dg::DGSEM)
```
Added   
```julia
# For quadrilateral elements, interpolation to plotting nodes involves applying a 1D interpolation
# operator to each line of nodes. This is equivalent to multiplying the multi-dimensional vector
# of node coordinates by a Kronecker product of the 1D interpolation operator (e.g., a
# multi-dimensional interpolation operator).
```

sloede replied:
Thanks a lot, this is very helpful. One remaining question though:
> multi-dimensional vector

For me, in Julia `vector`/`Vector` is always one-dimensional. Can we rephrase this to `array` or something similar?

jlchan replied:
Good point - here's a rewrite which is more clear about the multidimensional vector bit:

```julia
  # For quadrilateral elements, interpolation to plotting nodes involves applying a 1D interpolation
  # operator to each line of nodes. This is equivalent to multiplying the vector containing all node
  # node coordinates on an element by a Kronecker product of the 1D interpolation operator (e.g., a
  # multi-dimensional interpolation operator).
```

sloede replied:
:+1: Thanks, much clearer to me!

jlchan commented on src/visualization/plot_recipes.jl:
```diff
@@ -16,7 +16,7 @@ const TrixiODESolution = Union{ODESolution{T, N, uType, uType2, DType, tType, ra
 
 # This abstract type is used to derive PlotData types of different dimensions; but still allows to share some functions for them.
-abstract type AbstractPlotData end
+abstract type AbstractPlotData{Dim} end
```
I can never remember which one is the right one. Thanks!

jlchan commented on src/visualization/plot_recipes.jl:
```diff
@@ -293,7 +304,7 @@ end
 #
 # Note: This is an experimental feature and may be changed in future releases without notice.
-@recipe function f(pds::PlotDataSeries2D)
+RecipesBase.@recipe function f(pds::PlotDataSeries2D)
```
Yep

jlchan commented on test/test_visualization.jl:
```diff
@@ -263,6 +263,13 @@ isdir(outdir) && rm(outdir, recursive=true)
+  @testset "Makie visualization tests for UnstructuredMesh2D" begin
+    @test_nowarn_debug trixi_include(@__MODULE__, joinpath(examples_dir(), "unstructured_2d_dgsem", "elixir_euler_wall_bc.jl"))
+    @test_nowarn_debug Makie.plot(sol) # test heatmap plot
+    @test_nowarn_debug Trixi.trixi_plot(sol) # test interactive surface plot
```
Currently yes. It will work for `TreeMesh` and `StructuredMesh`, but is restricted to `UnstructuredMesh2D` for now. 

I want to extend to other mesh types, but I was hoping to do so after I add Plots.jl support in #755 so that I don't break any existing plotting functionality accidentally.

sloede replied:
:+1:

jlchan commented on test/Project.toml:
```diff
@@ -4,13 +4,15 @@ Downloads = "f43a241f-c20a-4ad4-852c-f6b1247861c6"
 ForwardDiff = "f6369f11-7733-5829-9624-2563aa707210"
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 MPI = "da04e1cc-30fd-572f-bb4f-1f8673147195"
+Makie = "ee78f7c6-11fb-53f2-987a-cfe4a2b5a57a"
```
I agree...it's like it's back to the early days of Plots.jl in terms of TTFP issues...

jlchan commented:
@sloede I think I've addressed most of your comments, and mentioned them in responses on this PR.  Would you let me know if I should clarify anything else in your previous set of review comments?

When that's done, I can re-request a review.

ranocha commented on docs/src/visualization.md:
````diff
@@ -432,7 +436,12 @@ Unlike the Plots.jl recipe, mesh plotting is controlled using the keyword argume
+julia> axes[1,1].title = "New title for subplot (1,1)"
+```
 
 Trixi also supports interactive surface plots using `trixi_plot`:
````
```suggestion
Trixi also supports interactive surface plots using `iplot`:
```
GitHub doesn't let me comment on it right now, but there is at least one additional `trixi_plot` below

jlchan replied:
Git ut

ranocha commented on src/visualization/makie_visualization.jl:
```diff
@@ -178,76 +184,91 @@ function generate_plotting_triangulation(pds::PlotDataSeries2D{<:UnstructuredPlo
 """
-    trixi_plot(sol::TrixiODESolution;
+    iplot(sol::TrixiODESolution;
                solution_variables=nothing, nvisnodes=5, variable_to_plot_in = 1)
```
```suggestion
    iplot(sol::TrixiODESolution;
          solution_variables=nothing, nvisnodes=5, variable_to_plot_in = 1)
```

ranocha commented on src/visualization/makie_visualization.jl:
```diff
@@ -178,76 +184,91 @@ function generate_plotting_triangulation(pds::PlotDataSeries2D{<:UnstructuredPlo
 """
-function trixi_plot(sol::TrixiODESolution;
-                    solution_variables=nothing, nvisnodes=5, variable_to_plot_in = 1)
+function iplot(sol::TrixiODESolution;
+               solution_variables=nothing, nvisnodes=default_nvisnodes(sol), variable_to_plot_in = 1)
```
What about using the signature `iplot(sol)` just as convenience version redirecting the work to `iplot(u, semi)` (or whatever signature you need)? This would make it much more convenient when working interactively, since we can copy and modify `u` instead of needing to create an ODE solution and modify it in place. Feel free to create an issue for this instead of implementing it here right now.

sloede replied:
I second this. That's what we also did for Plots.jl - the `sol` stuff is only syntactic sugar, redirecting to `u_ode, semi`, redirecting to the main implementation, `u, semi`.

ranocha replied:
Isn't the last one `u, mesh, equations, solver, cache` or something like this?

sloede replied:
You are - as usual - correct: https://github.com/trixi-framework/Trixi.jl/blob/9364f7fe5c5379c76e3a1fbeb5fcdd99dcd078cc/src/visualization/plot_recipes.jl#L111-L114

Thus, the sugar goes `sol` -> `u_ode, semi` -> `u, mesh, equations, solver, cache`. I think it would be nice if this were analogous for Makie.jl plotting.

ranocha commented on src/visualization/makie_visualization.jl:
```diff
@@ -312,13 +335,35 @@ Makie.plottype(::Trixi.PlotDataSeries2D{<:Trixi.UnstructuredPlotData2D}) = Trixi
+
+# allows for returning fig, axes = Makie.plot(...)
+function Base.iterate(fa::FigureAndAxes, state=1)
+  if state > 2
+    return nothing
+  elseif state==1
+    return (fa.fig, 2)
+  elseif state==2
+    return (fa.axes, 3)
+  end
```
```suggestion
  if state == 1
    return (fa.fig, 2)
  elseif state == 2
    return (fa.axes, 3)
  else
    return nothing
  end
```
This also handles `state < 1` (which might happen by accident if a user calls `iterate` explicitly with wrong `state`) and removes this branch, making it easier for the compiler.

ranocha reviewed (dismissed):
Thank you very much, @jlchan! I just have a couple of minor comments that basically don't block merging this :+1:

sloede commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,373 @@
+  r, s = reference_node_coordinates_2d(dg)
+
+  # reference plotting nodes
+  Vplot = plotting_interpolation_matrix(dg; nvisnodes=nvisnodes)
```
Or just `interpolation_matrix` - I think the `plotting` is implied. But yours works as well, thanks!

sloede commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,373 @@
+  return r, s
+end
+
+function transform_to_solution_variables!(u, solution_variables_, equations)
```
Thanks!

sloede commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,373 @@
+  Vplot = plotting_interpolation_matrix(dg; nvisnodes=nvisnodes)
+
+  # create triangulation for plotting nodes
+  rp, sp = (x->Vplot*x).((r, s)) # interpolate dg nodes to plotting nodes
```
Thanks!

sloede commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,373 @@
+  return UnstructuredPlotData2D(xplot, yplot, uplot, t, xfp, yfp, ufp, variable_names)
+end
+
+function generate_plotting_triangulation(pds::PlotDataSeries2D{<:UnstructuredPlotData2D};
```
That's much better, thanks a lot!

sloede reviewed (commented):
@jlchan I have unresolved a few comments which - I believe - were mistakenly marked as resolved. Please excuse if I forgot something or if I was too eager. Once the remaining conversations (there are a 2-3 open from myself, the rest from Hendrik), please request a re-review. This is then likely the final review & should just be a formality, since I think almost everything is ready.

jlchan commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,373 @@
+  else
+    sol_f = StructArrays.component(uf, variable_id)
+  end
+  xyz_wireframe = Makie.Point.(map(x->vec(vcat(x, fill(NaN, 1, size(x, 2)))), (xf, yf, sol_f))...)
```
```julia
  # This line separates solution lines on each edge by NaNs to ensure that they are rendered
  # separately. The coordinates `xf`, `yf` and the solution `sol_f`` are assumed to be a matrix
  # whose columns correspond to different elements. We add NaN separators by appending a row of
  # NaNs to this matrix. We also flatten (e.g., apply `vec` to) the result, as this speeds up
  # plotting.
```

jlchan commented on src/visualization/makie_visualization.jl:
```diff
@@ -0,0 +1,373 @@
+  return plotting_mesh
+end
+
+function generate_plotting_wireframe(pds::PlotDataSeries2D{<:UnstructuredPlotData2D};
```
```julia
# Returns a list of `Makie.Point`s which can be used to plot the mesh, or a solution "wireframe"
# (e.g., a plot of the mesh lines but with the z-coordinate equal to the value of the solution).
function mesh_plotting_wireframe(pds::PlotDataSeries2D{<:UnstructuredPlotData2D};
                                     set_z_coordinate_zero = false)
```

jlchan commented:
Alright - I think I've addressed most of the review comments. Should be ready for another round of reviews.

sloede reviewed (approved):
Dear Makie, welcome to the Trixi experience!

LGTM, great work @jlchan and thanks for putting in all the work!

ranocha commented:
https://github.com/trixi-framework/Trixi.jl/pull/613/checks?check_run_id=3346909423#step:6:4661

sloede commented:
> https://github.com/trixi-framework/Trixi.jl/pull/613/checks?check_run_id=3346909423#step:6:4661

Maybe fixed in https://github.com/trixi-framework/Trixi.jl/pull/613/commits/b631083cf59e3adaa3208294b4cdfc2e071c04f5

ranocha commented:
> ┌ Warning:     OpenGL/GLFW wasn't loaded correctly or couldn't be initialized.
│     This likely means, you're on a headless server without having OpenGL support setup correctly.
│     Have a look at the troubleshooting section in the readme:
│     https://github.com/JuliaPlots/Makie.jl/tree/master/GLMakie#troubleshooting-opengl.
└ @ GLMakie ~/.julia/packages/GLMakie/rXGL8/src/gl_backend.jl:4
ERROR: LoadError: LoadError: InitError: Exception[GLFW.GLFWError(GLFW.PLATFORM_ERROR, "X11: The DISPLAY environment variable is missing"), ErrorException("glfwInit failed")]
Stacktrace:
  [1] __init__()
    @ GLFW ~/.julia/packages/GLFW/BWxfF/src/GLFW.jl:35

See https://github.com/trixi-framework/Trixi.jl/pull/613/checks?check_run_id=3347951169#step:6:4659

sloede commented:
> > ┌ Warning:     OpenGL/GLFW wasn't loaded correctly or couldn't be initialized.
> > │     This likely means, you're on a headless server without having OpenGL support setup correctly.
> > │     Have a look at the troubleshooting section in the readme:
> > │     https://github.com/JuliaPlots/Makie.jl/tree/master/GLMakie#troubleshooting-opengl.
> > └ @ GLMakie ~/.julia/packages/GLMakie/rXGL8/src/gl_backend.jl:4
> > ERROR: LoadError: LoadError: InitError: Exception[GLFW.GLFWError(GLFW.PLATFORM_ERROR, "X11: The DISPLAY environment variable is missing"), ErrorException("glfwInit failed")]
> > Stacktrace:
> > [1] **init**()
> > @ GLFW ~/.julia/packages/GLFW/BWxfF/src/GLFW.jl:35
> 
> See https://github.com/trixi-framework/Trixi.jl/pull/613/checks?check_run_id=3347951169#step:6:4659

Maye we can use something from [here](http://makie.juliaplots.org/stable/remote.html#GLMakie-in-CI)?
https://github.com/JuliaPlots/GLMakie.jl/blob/8954fc34354a09ceb11159a8e8e35429c05a710f/.github/workflows/ci.yml#L41-L42

ranocha commented:
Or CairoMakie?

> For CairoMakie there shouldn't be any difference in using it on a remote or locally.

See http://makie.juliaplots.org/stable/remote.html#CairoMakie

sloede commented:
> Or CairoMakie?

Unfortunately, CairoMakie is only non interactive, at least as far as I can tell from [here](https://makie.juliaplots.org/stable/#Makie-Ecosystem). I don't know how this works out when trying something "forbidden", but I think chances are high that this will not allow us to test `Trixi.iplot`. @jlchan have you tried this yourself?

jlchan commented:
@sloede let me try CairoMakie out

jlchan commented:
@sloede @ranocha CairoMakie does let us test `iplot` - it just creates a png of the interactive window. Could we use that instead?

jlchan commented:
Great - looks like the `misc` test passes with CairoMakie.

sloede commented:
> Great - looks like the `misc` test passes with CairoMakie.

Awesome! Let's hope that Codecov does not complain either!
